src/libraries/System.Private.CoreLib/src/System/Array.cs

```diff
@@ -207,6 +207,74 @@ public static Array CreateInstance(Type elementType, params long[] lengths)
             return CreateInstance(elementType, intLengths);
         }
 
+        public static unsafe Array CreateInstanceFromArrayType(Type arrayType, int length)
```

Comment:
I added `DynamicallyAccessedMembers` in a first try, and caused other places accessing arrays starting to report trimmer warning. This is an actual blocker to resolve.

Reply:
That's because of the implementation goes from array type to element type, and then re-creates the array from element type again. It causes the warnings and it also makes the API much slower than it needs to be.

Reply:
The actual problem is quite complicated and I can't explain its detail without my machine. In short, something in `System.Xml` is passing array type to method annotated with `DynamicallyAccessedMembers.All`, and thus the static members are considered access, then the `DynamicallyAccessedMembers.Constructors` triggers warning. It may be resolved in some trimmer change, I haven't confirmed yet.

Reply:
> That's because of the implementation goes from array type to element type, and then re-creates the array from element type again. It causes the warnings and it also makes the API much slower than it needs to be.

@jkotas  I understand that. What can I use instead of `InternalCreate`? What method does the compiler use?

Reply:
I think it is impossible to use pathed `arrayType` in all cases 😥. I think in 99,9% of cases we can use `arrayType` to create an array but sometimes not. For example `arrayType` can be a type for a one-dimensional array and the user want to create a multi-dimensional array. So method should find another array class for a multi-dimensional array and create it. This method can change array dimensions, so different array classes could be created.

Reply:
> I think it is impossible to use pathed `arrayType` in all cases 😥. I think in 99,9% of cases we can use `arrayType` to create an array but sometimes not. For example `arrayType` can be a type for a one-dimensional array and the user want to create a multi-dimensional array. So method should find another array class for a multi-dimensional array and create it. This method can change array dimensions, so different array classes could be created.

Somehow Github made this comment appear twice but this is my response to the other instance: https://github.com/dotnet/runtime/pull/88620#discussion_r1259420229. We don't need that behavior. If type doesn't match dimensions, it's an error.